File: engine/model/src/main/scala/client/Ring.scala
package client

class Ring(private var _ringId: Int, private var _partnerRingId: Int, private var _partnerId: Int, private var _itemId: Int, private var _partnerName: String) extends Comparable[Ring] {
  def ringId: Int = _ringId

  def partnerRingId: Int = _partnerRingId

  def partnerId: Int = _partnerId

  def itemId: Int = _itemId

  def partnerName: String = _partnerName

  private var equipped: Boolean = false

  def isEquipped: Boolean = {
    equipped
  }

  def equip(): Unit = {
    equipped = true
  }

  def unequip(): Unit = {
    equipped = false
  }

  override def hashCode(): Int = {
    var hash: Int = 5
    hash = 53 * hash + _ringId
    hash
  }

  override def equals(obj: Any): Boolean = {
    obj match {
      case ring: Ring =>
        ring._ringId == _ringId
      case _ =>
        false
    }
  }

  override def compareTo(o: Ring): Int = {
    if (_ringId < o._ringId) {
      -1
    } else if (_ringId == o._ringId) {
      0
    } else {
      1
    }
  }
}

File: engine/engine-model/src/main/java/client/Ring.java
package client;

import java.util.Objects;

public record Ring(int ringId, int partnerRingId, int partnerId, int itemId, String partnerName,
                   boolean equipped) implements Comparable<Ring> {
   public Ring(int ringId, int partnerRingId, int partnerId, int itemId, String partnerName) {
      this(ringId, partnerRingId, partnerId, itemId, partnerName, false);
   }

   public Ring equip() {
      return new Ring(ringId(), partnerRingId(), partnerId(), itemId(), partnerName(), true);
   }

   public Ring unequip() {
      return new Ring(ringId(), partnerRingId(), partnerId(), itemId(), partnerName(), false);
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) {
         return true;
      }
      if (o == null || getClass() != o.getClass()) {
         return false;
      }
      Ring ring = (Ring) o;
      return Objects.equals(ringId, ring.ringId());
   }

   @Override
   public int hashCode() {
      return Objects.hash(ringId);
   }

   @Override
   public int compareTo(Ring o) {
      if (ringId < o.ringId()) {
         return -1;
      } else if (ringId == o.ringId()) {
         return 0;
      }
      return 1;
   }
}
